metadata-io/src/main/java/com/linkedin/metadata/entity/UrnValidator.java

```diff
@@ -0,0 +1,86 @@
+import com.linkedin.data.schema.DataSchema.Type;
+import com.linkedin.data.schema.NamedDataSchema;
+import com.linkedin.data.schema.PathSpec;
+import com.linkedin.data.schema.validator.Validator;
+import com.linkedin.data.schema.validator.ValidatorContext;
+import com.linkedin.data.template.RecordTemplate;
+import com.linkedin.metadata.models.EntitySpec;
+import com.linkedin.metadata.models.RelationshipFieldSpec;
+import com.linkedin.metadata.models.registry.EntityRegistry;
+import com.linkedin.metadata.utils.EntityKeyUtils;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.Setter;
+
+
+public class UrnValidator implements Validator {
+
+  private final EntityRegistry _entityRegistry;
+  @Setter
+  private EntitySpec currentEntitySpec = null;
+
+  public UrnValidator(EntityRegistry entityRegistry) {
+    _entityRegistry = entityRegistry;
+  }
+
+  public void validate(ValidatorContext context) {
+    if (currentEntitySpec == null) {
+      throw new IllegalStateException("Current entity spec must be set");
+    }
+    if (Type.TYPEREF.equals(context.dataElement().getSchema().getType()) && ((NamedDataSchema) context.dataElement()
+        .getSchema()).getName().endsWith("Urn")) {
+      try {
+        // Validate Urn matches field type and that it generates a valid key
+        String urnStr = (String) context.dataElement().getValue();
+        Urn urn = Urn.createFromString(urnStr);
+        EntitySpec entitySpec = _entityRegistry.getEntitySpec(urn.getEntityType());
+        RecordTemplate entityKey = EntityKeyUtils.convertUrnToEntityKey(urn,
+            entitySpec.getKeyAspectSpec().getPegasusSchema());
+        NamedDataSchema namedDataSchema = ((NamedDataSchema) context.dataElement().getSchema());
+        Class<? extends Urn> urnClass;
+        try {
+          String schemaName = ((Map<String, String>) namedDataSchema.getProperties().get("java")).get("class");
+          urnClass = (Class<? extends Urn>) Class.forName(schemaName);
+          urnClass.getDeclaredMethod("createFromString", String.class).invoke(null, urnStr);
+        } catch (ClassNotFoundException | ClassCastException | NoSuchMethodException e) {
+          throw new IllegalArgumentException("Unrecognized Urn class: " + namedDataSchema.getName(), e);
+        } catch (InvocationTargetException | IllegalAccessException e) {
+          throw new IllegalArgumentException("Unable to instantiate urn type: " + namedDataSchema.getName() + " with urn: " + urnStr, e);
+        }
+
+        // Validate generic Urn is valid entity type for relationship destination
+        PathSpec fieldPath = context.dataElement().getSchemaPathSpec();
+        List<RelationshipFieldSpec> relationshipSpecs = currentEntitySpec.getRelationshipFieldSpecs().stream().filter(relationshipFieldSpec ->
+            relationshipFieldSpec.getPath().equals(fieldPath))
```

Comment:
Very nice! Did you validate this against these cases:

1. Urn as field 
2. Urn inside array field
3. Urn inside a complex field with Relationship "override" (e.g. specifying the path override for an annotation)